parse.py

```diff
@@ -0,0 +1,292 @@
+	If no root is provided, root will be set to html5.Body().
+
+	The HTML elements are parsed for notations of kind [name]="ident", making
+	the corresponding instance available to the widget as widget.ident in the
+	Python code.
+
+	Example:
+
+	```python
+	import html5
+
+	div = html5.Div()
+	html5.parse.fromHTML('''
+		<div>Yeah!
+			<a href="hello world" [name]="myLink" class="trullman bernd" disabled>
+			hah ala malla" bababtschga"
+			<img src="/static/images/icon_home.svg" style="background-color: red;"/>st
+			<em>ah</em>ralla <i>malla tralla</i> da
+			</a>lala
+		</div>''', div)
+
+	div.myLink.appendChild("appended!")
+	```
+	"""
+
+	def scanWhite(l):
+		"""
+		Scan and return whitespace.
+		"""
+
+		ret = ""
+		while l and l[0] in string.whitespace:
+			ret += l.pop(0)
+
+		return ret
+
+	def scanWord(l):
+		"""
+		Scan and return a word.
+		"""
+
+		ret = ""
+		while l and l[0] not in string.whitespace + "<>=\"'":
+			ret += l.pop(0)
+
+		return ret
+
+	global _tags
+	stack = []
+
+	# Obtain tag descriptions
+	if _tags is None:
+		_tags = _buildDescription()
+
+	# Handle defaults
+	if appendTo is None:
+		appendTo = html5.Body()
+
+	if bindTo is None:
+		bindTo = appendTo
+
+	# Prepare stack and input
+	stack.append((appendTo, None))
+	html = [ch for ch in html]
+
+	# Parse
+	while html:
+		tag = None
+		text = ""
+
+		# ugly...
+		while stack and stack[-1][1] in ["br", "input", "img"]:
+			stack.pop()
+
+		if not stack:
+			break
+
+		parent = stack[-1][0]
+
+		while html:
+			#print("html", html)
+			#print(stack)
+
+			ch = html.pop(0)
+
+			# Comment
+			if html and ch == "<" and "".join(html[:3]) == "!--":
+				html = html[3:]
+				while html and "".join(html[:3]) != "-->":
+					html.pop(0)
+
+				html = html[3:]
+
+			# Opening tag
+			elif html and ch == "<" and html[0] != "/":
+				tag = scanWord(html)
+				if tag.lower() in _tags:
+					break
+
+				text += ch + tag
+
+			# Closing tag
+			elif html and stack[-1][1] and ch == "<" and html[0] == "/":
+				junk = ch
+				junk += html.pop(0)
+
+				tag = scanWord(html)
+				junk += tag
+
+				#print("/", tag.lower(), stack[-1][1].lower())
+				if stack[-1][1].lower() == tag.lower():
+					junk += scanWhite(html)
+					if html and html[0] == ">":
+						html.pop(0)
+						stack.pop()
+						tag = None
+						break
+
+				text += junk
+				tag = None
+
+			else:
+				text += ch
+
+		# Append plain text (if not only whitespace)
+		if (text
+			and ((len(text) == 1 and text in ["\t "])
+		        or not all([ch in string.whitespace for ch in text]))):
+
+			#print("text", text)
+			parent.appendChild(html5.TextNode(_convertEncodedText(text)))
+
+		# Create tag
+		if tag:
+			wdg = _tags[tag][0]()
+
+			parent.appendChild(wdg)
+			stack.append((wdg, tag))
+
+			#print("tag", tag)
+
+			while html:
+				scanWhite(html)
+				if not html:
+					break
+
+				# End of tag >
+				if html[0] == ">":
+					html.pop(0)
+					break
+
+				# Closing tag at end />
```

Comment:
It's hard to follow the flow here, why are there two "closing tag at end />" sections?

Reply:
As explained above.